### PR TITLE
[6X] Add GUC gp_workfile_compression_overhead_limit for ZSTD buffer

### PR DESCRIPTION
**Pull request by yaowangm**
It's a back port of https://github.com/greenplum-db/gpdb/pull/16243. However, besides the port there are still some essential changes:

1. Removed the statistic info of max/min file size, and used average file size instead;
2. Included the work file number of not only hash table (for inner table), but also outer table;
3. Updated the test case zlib to compare work file number and compressed work file number to avoid checking accurate file number on different envs (planner vs. orca, 6X vs. 7X)

The changes should be forward-ported to 7X again after the PR is merged to keep the behaviors consistent.

### CODE REVIEW

**Comment by yaowangm** [comment]
When I ran CI pipeline the compilation failed due to the error:
```
buffile.c: In function 'BufFileDumpCompressedBuffer':
buffile.c:1127:27: error: implicit declaration of function 'ZSTD_sizeof_CStream'; did you mean 'ZSTD_initCStream'? [-Werror=implicit-function-declaration]
  compressed_buffer_size = ZSTD_sizeof_CStream(file->zstd_context->cctx);
                           ^~~~~~~~~~~~~~~~~~~
                           ZSTD_initCStream
```
The root cause is `ZSTD_sizeof_CStream` is marked as ZSTD_STATIC_LINKING_ONLY in zstd.h on 6X. However it's generally available on 7X because the zstd lib version is newer on 7X. 

So the question is, Is it possible to upgrade the zstd lib on 6X? If not, we may use kind of estimation algorithm to calculate the buffer size used by zstd. As we watched,  the buffer size is relative stable so theoretically we can do the estimation... @adam8157

**Comment by yaowangm** [comment]
Some investigations:

I did a couple of experiments, and when the temp files are small (< 1MB), the buffer size used by zstd lib per file is pretty stable (about 1.3 MB). See the two examples:
```
 (seg1)   Work file set: 126 files (126 compressed), avg file size 68336, compression buffer size 1714
23630 bytes
```
The avg buffer size is ~1.3 MB per file when avg file size is ~68KB.
```
 (seg1)   Work file set: 6 files (6 compressed), avg file size 957038, compression buffer size 8163030
 bytes
```
The avg buffer size is ~1.3 MB per file when avg file size is ~957KB.

It seems we may use 1.3MB as an estimated value for buffer size and void the using of `ZSTD_sizeof_CStream`. When the file size is larger, the buffer size might increase. But I have not hit the scenario with a large temp file. When I enlarged the `statement_mem` for a larger hash table (and larger temp file), in fact no spilled file was created. Maybe more investigation is needed. Anyway, it's perfect if we can upgrade zstd lib and use the API directly.

**Comment by yaowangm** [comment]
I ran a big query for more than one hour and generated bigger file size:
```
testdb=# explain analyze
testdb-# with B as (select distinct a+1 as a,b,c from test_memory)
testdb-# ,C as (select distinct a+2 as a,b,c from test_memory)
testdb-# ,D as (select a+3 as a,b,c from test_memory)
testdb-# ,E as (select a+4 as a,b,c from test_memory)
testdb-# ,F as (select (a+5)::text as a,b,c from test_memory)
testdb-# select count(*) from test_memory A
testdb-# inner join  B on A.a = B.a
testdb-# inner join  C on A.a = C.a
testdb-# inner join  D on A.a = D.a
testdb-# inner join  E on A.a = E.a
testdb-# inner join  F on A.a::text = F.a ;

                         QUERY PLAN

------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------
-------------------------------------------------------------
 Aggregate  (cost=1984267692086.20..1984267692086.21 rows=1 width=8) (actual time=3839672.912..3839672
.912 rows=1 loops=1)
   ->  Gather Motion 3:1  (slice8; segments: 3)  (cost=1984267692086.14..1984267692086.19 rows=1 width
=8) (actual time=3832820.548..3839672.896 rows=3 loops=1)
         ->  Aggregate  (cost=1984267692086.14..1984267692086.15 rows=1 width=8) (actual time=3835501.
512..3835501.512 rows=1 loops=1)
               ->  Merge Join  (cost=20167724903.09..1905703823097.18 rows=10475182531861 width=0) (ac
tual time=3239542.447..3818904.572 rows=243559327 loops=1)
                     Merge Cond: (f.a = ((a.a)::text))
                     ->  Sort  (cost=26578550.50..26853601.12 rows=36673416 width=32) (actual time=184
077.677..220908.198 rows=36672736 loops=1)
                           Sort Key: f.a
                           Sort Method:  external merge  Disk: 1720192kB
                           ->  Subquery Scan on f  (cost=0.00..2200404.96 rows=36673416 width=32) (act
ual time=0.035..18670.546 rows=36672736 loops=1)
                                 ->  Seq Scan on test_memory  (cost=0.00..2126716.34 rows=36673416 wid
th=25) (actual time=0.034..14145.360 rows=36672736 loops=1)
                     ->  Materialize  (cost=10141146352.59..10145430865.87 rows=285634219 width=4) (ac
tual time=1527732.366..1703831.297 rows=258550039 loops=2)
                           ->  Sort  (cost=10141146352.59..10143288609.23 rows=285634219 width=4) (act
ual time=3055464.727..3339186.216 rows=410159898 loops=1)
                                 Sort Key: ((a.a)::text)
                                 Sort Method:  external merge  Disk: 28782912kB
                                 ->  Result  (cost=5774173962.49..9927812862.89 rows=285634219 width=4
) (actual time=460112.308..697166.952 rows=410159898 loops=1)
                                       ->  Broadcast Motion 3:3  (slice7; segments: 3)  (cost=57741739
62.49..9927812862.89 rows=285634219 width=4) (actual time=460112.302..568372.507 rows=410159898 loops=
1)
                                             ->  Hash Join  (cost=5774173962.49..9916387494.13 rows=95
211407 width=4) (actual time=460563.902..570667.230 rows=136875177 loops=1)
                                                   Hash Cond: (e.a = a.a)
                                                   Extra Text: (seg0)   Initial batch 0:
 (seg0)     Wrote 2245193K bytes to inner workfile.
 (seg0)     Wrote 422979K bytes to outer workfile.
 (seg0)   Initial batches 1..63:
 (seg0)     Read 2245193K bytes from inner workfile: 35638K avg x 63 nonempty batches, 37821K max.
 (seg0)     Wrote 162381K bytes to inner workfile: 18043K avg x 9 overflowing batches, 20093K max.
 (seg0)     Read 422979K bytes from outer workfile: 6714K avg x 63 nonempty batches, 6737K max.
 (seg0)     Wrote 30212K bytes to outer workfile: 3357K avg x 9 overflowing batches, 3364K max.
 (seg0)   Secondary Overflow batches 64..127:
 (seg0)     Read 162381K bytes from inner workfile: 18043K avg x 9 nonempty batches, 20093K max.
 (seg0)     Read 30212K bytes from outer workfile: 3357K avg x 9 nonempty batches, 3364K max.
 (seg0)   Work file set: 126 files (126 compressed), avg file size 4628524, compression buffer size 17
1423630 bytes
 (seg0)   Hash chain length 3.8 avg, 80016 max, using 21770163 of 38273024 buckets.  Skipped 55 empty
batches.Initial batch 0:

                                                   ->  Redistribute Motion 3:3  (slice1; segments: 3)
 (cost=0.00..4400809.92 rows=36673416 width=4) (actual time=0.119..24666.955 rows=36679988 loops=1)
                                                         Hash Key: e.a
                                                         ->  Subquery Scan on e  (cost=0.00..2200404.9
6 rows=36673416 width=4) (actual time=0.027..11846.532 rows=36672736 loops=1)
                                                               ->  Seq Scan on test_memory test_memory
_1  (cost=0.00..1576615.10 rows=36673416 width=25) (actual time=0.026..7851.807 rows=36672736 loops=1)
                                                   ->  Hash  (cost=5771741657.92..5771741657.92 rows=5
9090842 width=16) (actual time=460106.471..460106.471 rows=83413140 loops=1)
                                                         ->  Hash Join  (cost=3199113535.56..577174165
7.92 rows=59090842 width=16) (actual time=357257.147..435125.817 rows=83413140 loops=1)
                                                               Hash Cond: (d.a = a.a)
                                                               Extra Text: (seg2)   Initial batch 0:
 (seg2)     Wrote 1089613K bytes to inner workfile.
 (seg2)     Wrote 422991K bytes to outer workfile.
 (seg2)   Initial batches 1..63:
 (seg2)     Read 1089613K bytes from inner workfile: 17296K avg x 63 nonempty batches, 17426K max.
 (seg2)     Read 422991K bytes from outer workfile: 6715K avg x 63 nonempty batches, 6747K max.
 (seg2)   Work file set: 126 files (126 compressed), avg file size 4958234, compression buffer size 17
1423630 bytes
 (seg2)   Hash chain length 3.9 avg, 50 max, using 14473556 of 16777216 buckets.Initial batch 0:

                                                               ->  Redistribute Motion 3:3  (slice2; s
egments: 3)  (cost=0.00..4400809.92 rows=36673416 width=4) (actual time=0.115..12422.404 rows=36677278
 loops=1)
                                                                     Hash Key: d.a
                                                                     ->  Subquery Scan on d  (cost=0.0
0..2200404.96 rows=36673416 width=4) (actual time=0.021..11413.855 rows=36672736 loops=1)
                                                                           ->  Seq Scan on test_memory
 test_memory_2  (cost=0.00..1576615.10 rows=36673416 width=25) (actual time=0.021..7715.512 rows=36672
736 loops=1)
                                                               ->  Hash  (cost=3197603979.42..31976039
79.42 rows=36673417 width=12) (actual time=341439.840..341439.840 rows=56679022 loops=1)
                                                                     ->  Hash Join  (cost=1600935457.3
6..3197603979.42 rows=36673417 width=12) (actual time=212725.964..324008.784 rows=56679022 loops=1)
                                                                           Hash Cond: (c.a = a.a)
                                                                           Extra Text: (seg2)   Initia
l batch 0:
 (seg2)     Wrote 839878K bytes to inner workfile.
 (seg2)     Wrote 1563327K bytes to outer workfile.
 (seg2)   Initial batches 1..63:
 (seg2)     Read 416492K bytes from inner workfile: 6611K avg x 63 nonempty batches, 6667K max.
 (seg2)     Read 775266K bytes from outer workfile: 12306K avg x 63 nonempty batches, 12390K max.
 (seg2)   Overflow batches 64..127:
 (seg2)     Read 423386K bytes from inner workfile: 6616K avg x 64 nonempty batches, 6683K max.
 (seg2)     Read 788062K bytes from outer workfile: 12314K avg x 64 nonempty batches, 12410K max.
 (seg2)   Work file set: 254 files (254 compressed), avg file size 4445795, compression buffer size 34
5568270 bytes
 (seg2)   Hash chain length 2.1 avg, 21 max, using 21102297 of 33554432 buckets.36671800 groups total
in 160 batches; 33 overflows; 73343600 spill groups.

                                                                           ->  Redistribute Motion 3:3
  (slice4; segments: 3)  (cost=0.00..2731264.84 rows=22760541 width=4) (actual time=0.049..52013.283 r
ows=36669987 loops=1)
                                                                                 Hash Key: c.a
                                                                                 ->  Subquery Scan on
c  (cost=0.00..1365632.42 rows=22760541 width=4) (actual time=78232.420..132547.189 rows=36671800 loop
s=1)
                                                                                       ->  HashAggrega
te  (cost=6507980.36..7674235.78 rows=22760541 width=44) (actual time=78232.419..129582.438 rows=36671
800 loops=1)
                                                                                             Group Key
: ((test_memory_3.a + 2)), test_memory_3.b, test_memory_3.c
                                                                                             Extra Tex
t: (seg1)   36671800 groups total in 160 batches; 33 overflows; 73343600 spill groups.
 (seg1)   Hash chain length 4.5 avg, 20 max, using 10429629 of 10551296 buckets; total 33 expansions.

                                                                                             ->  Redis
tribute Motion 3:3  (slice3; segments: 3)  (cost=2401766.96..4829612.78 rows=22760541 width=44) (actua
l time=527.027..44145.862 rows=36671800 loops=1)
                                                                                                   Has
h Key: ((test_memory_3.a + 2)), test_memory_3.b, test_memory_3.c
                                                                                                   ->
 HashAggregate  (cost=2401766.96..3463980.36 rows=22760541 width=44) (actual time=524.912..34187.206 r
ows=36669984 loops=1)

   Group Key: (test_memory_3.a + 2), test_memory_3.b, test_memory_3.c

   Extra Text: (seg2)   Hash chain length 4.9 avg, 19 max, using 7411405 of 7471104 buckets; total 1 e
xpansions.


   ->  Seq Scan on test_memory test_memory_3  (cost=0.00..1576615.10 rows=36673416 width=25) (actual t
ime=0.021..9131.762 rows=36672736 loops=1)
                                                                           ->  Hash  (cost=1599452762.
24..1599452762.24 rows=36673417 width=8) (actual time=212719.110..212719.110 rows=43339663 loops=1)
                                                                                 ->  Hash Join  (cost=
2784259.58..1599452762.24 rows=36673417 width=8) (actual time=88295.730..198047.350 rows=43339663 loop
s=1)
                                                                                       Hash Cond: (b.a
 = a.a)
                                                                                       Extra Text: (se
g1)   Initial batch 0:
 (seg1)     Wrote 423055K bytes to inner workfile.
 (seg1)     Wrote 1550718K bytes to outer workfile.
 (seg1)   Initial batches 1..63:
 (seg1)     Read 423055K bytes from inner workfile: 6716K avg x 63 nonempty batches, 6828K max.
 (seg1)     Read 1550718K bytes from outer workfile: 24615K avg x 63 nonempty batches, 24702K max.
 (seg1)   Work file set: 126 files (126 compressed), avg file size 7624239, compression buffer size 17
1423630 bytes
 (seg1)   Hash chain length 2.5 avg, 10003 max, using 14475054 of 16777216 buckets.36680034 groups tot
al in 160 batches; 33 overflows; 73360068 spill groups.

                                                                                       ->  Redistribut
e Motion 3:3  (slice6; segments: 3)  (cost=0.00..2731264.84 rows=22760541 width=4) (actual time=66866.
129..119255.989 rows=36669986 loops=1)
                                                                                             Hash Key:
 b.a
                                                                                             ->  Subqu
ery Scan on b  (cost=0.00..1365632.42 rows=22760541 width=4) (actual time=88312.336..142935.920 rows=3
6680034 loops=1)
                                                                                                   ->
 HashAggregate  (cost=6507980.36..7674235.78 rows=22760541 width=44) (actual time=88312.334..140047.80
7 rows=36680034 loops=1)

   Group Key: ((test_memory_4.a + 1)), test_memory_4.b, test_memory_4.c

   Extra Text: (seg2)   36680034 groups total in 160 batches; 33 overflows; 73360068 spill groups.
 (seg2)   Hash chain length 4.5 avg, 20 max, using 10429507 of 10551296 buckets; total 33 expansions.


   ->  Redistribute Motion 3:3  (slice5; segments: 3)  (cost=2401766.96..4829612.78 rows=22760541 widt
h=44) (actual time=524.161..54475.429 rows=36680034 loops=1)

         Hash Key: ((test_memory_4.a + 1)), test_memory_4.b, test_memory_4.c

         ->  HashAggregate  (cost=2401766.96..3463980.36 rows=22760541 width=44) (actual time=525.870.
.33699.890 rows=36669984 loops=1)

               Group Key: (test_memory_4.a + 1), test_memory_4.b, test_memory_4.c

               Extra Text: (seg2)   Hash chain length 4.9 avg, 20 max, using 7411492 of 7471104 bucket
s; total 1 expansions.


               ->  Seq Scan on test_memory test_memory_4  (cost=0.00..1576615.10 rows=36673416 width=2
5) (actual time=0.022..9116.789 rows=36672736 loops=1)
                                                                                       ->  Hash  (cost
=1301564.48..1301564.48 rows=36673416 width=4) (actual time=21544.749..21544.749 rows=36672736 loops=1
)
                                                                                             ->  Seq S
can on test_memory a  (cost=0.00..1301564.48 rows=36673416 width=4) (actual time=0.021..9840.894 rows=
36672736 loops=1)
 Planning time: 5.941 ms
   (slice0)    Executor memory: 544K bytes.
   (slice1)    Executor memory: 62K bytes avg x 3 workers, 62K bytes max (seg0).
   (slice2)    Executor memory: 62K bytes avg x 3 workers, 62K bytes max (seg0).
   (slice3)    Executor memory: 21367K bytes avg x 3 workers, 21367K bytes max (seg0).
 * (slice4)    Executor memory: 26491K bytes avg x 3 workers, 26491K bytes max (seg1).  Work_mem: 2122
1K bytes max, 2299157K bytes wanted.
   (slice5)    Executor memory: 21367K bytes avg x 3 workers, 21367K bytes max (seg0).
 * (slice6)    Executor memory: 26491K bytes avg x 3 workers, 26491K bytes max (seg0).  Work_mem: 2122
1K bytes max, 2299797K bytes wanted.
 * (slice7)    Executor memory: 186300K bytes avg x 3 workers, 186599K bytes max (seg2).  Work_mem: 53
001K bytes max, 3258326K bytes wanted.
 * (slice8)    Executor memory: 108325K bytes avg x 3 workers, 108341K bytes max (seg2).  Work_mem: 54
490K bytes max, 35241474K bytes wanted.
 Memory used:  128000kB
 Memory wanted:  387658905kB
 Optimizer: Postgres query optimizer
 Execution time: 3840501.170 ms
(125 rows)

Time: 3840509.408 ms
```
When the avg file size is more than 4MB, the compression buffer size is still about 1.3MB. Although larger file size is possible form some corner case, I suppose it's good enough to use a fixed size of compression buffer size for calculation. I will change the code accordingly.

**Comment by adam8157** [comment]
1, the new changes better go into 7X first.
2, `ZSTD_sizeof_CStream` is marked as `ZSTD_STATIC_LINKING_ONLY` likely because it's not stable, so Zstd asks to static link it to have a fixed behavior.

And we need to get some feedback about upgrading the zstd.

**Comment by yaowangm** [comment]
2, `ZSTD_sizeof_CStream` is marked as `ZSTD_STATIC_LINKING_ONLY` likely because it's not stable, so Zstd asks to static link it to have a fixed behavior.
> 
> And we need to get some feedback about upgrading the zstd.

@adam8157 As I mentioned, `ZSTD_sizeof_CStream` is marked as `ZSTD_STATIC_LINKING_ONLY` **only on GPDB6** (zstd ver 1.3.7). It's generally available on GPDB7 (zstd ver 1.4.4). So the stability problem has been fixed. The question for GPDB is just whether we will upgrade zstd version on 6X to 1.4.4 in future.

**Comment by adam8157** [comment]
Exactly, we need PM feedback on upgrading zstd on 6X versions.

File: src/include/executor/hashjoin.h [hunk 1]
```diff
@@ -208,6 +208,12 @@ typedef struct HashJoinTableData
 
     HashJoinState * hjstate; /* reference to the enclosing HashJoinState */
     bool first_pass; /* Is this the first pass (pre-rescan) */
+
+	/* Statistic info of work file set, copied from work_set */
+	uint32      workset_num_files;
+	uint32      workset_num_files_compressed;
+	uint64      workset_avg_file_size;
+	uint64      workset_compression_buf_total;
```
Comment (adam8157) [inline src/include/executor/hashjoin.h:216]: `HashJoinTableData` is not in other structures, adding elements to the end is ABI safe I think.

File: src/include/utils/workfile_mgr.h [hunk 2]
```diff
@@ -87,6 +87,12 @@ typedef struct workfile_set
 
 	/* Used to track workfile_set created in current process */
 	dlist_node	local_node;
+
+	/* Total memory usage by compression buffer */
+	uint64      compression_buf_total;
+
+	/* Number of compressed work files */
+	uint32      num_files_compressed;
```
Comment (adam8157) [inline src/include/utils/workfile_mgr.h:95]: `workfile_set` is only in other structures as the form of a pointer, adding elements to the end is ABI safe too, I think.

**Review by adam8157 — APPROVED** [review]
I like the hardcoding than upgrading zstd.

**Pull request merged**